### PR TITLE
NAS-119900 / Python bindings for SES (sg_ses from sg3-utils)

### PR DESCRIPTION
**Pull request by umsaleem**
Enclosure plugin on SCALE sub-processes too much and uses sg_ses command to query the enclosure information.

This PR provides a Cython plugin that can be used without `subprocess.run()` to query enclosure information. Support for retrieving SES Configuration (cf), Enclosure Status (es), Element Descriptor (ed)  diagnostic pages is added. Output is similar to what `sg_ses` command returns.

`status()` returns a python dictionary with information extracted from the pages.

```
>>> from libsg3 import ses
>>> dev = ses.EnclosureDevice("/dev/bsg/3:0:0:0")
>>> dev.status()
{'id': '3000000000000001', 'name': 'AHCI SGPIO Enclosure 2.00', 'status': {'OK'}, 'elements': {0: {'type': 23, 'descriptor': 'Drive Slots', 'status': [0, 0, 0, 0]}, 1: {'type': 23, 'descriptor': 'Slot 00', 'status': [1, 0, 0, 0]}, 2: {'type': 23, 'descriptor': 'Slot 01', 'status': [1, 0, 0, 0]}}}
```
Configuration Diagnostic page:
```
>>> print(dev.get_configuration())
  AHCI      SGPIO Enclosure   2.00
Configuration diagnostic page:
  number of secondary subenclosures: 0
  generation code: 0x0
  enclosure descriptor list
    Subenclosure identifier: 0 [primary]
      relative ES process id: 1, number of ES processes: 1
      number of type descriptor headers: 1
      enclosure logical identifier (hex): 3000000000000001
      enclosure vendor: AHCI      product: SGPIO Enclosure   rev: 2.00
  type descriptor header and text list
    Element type: Array device slot, subenclosure id: 0
      number of possible elements: 2
      text: Drive Slots
```
Element Descriptor page:
```
>>> print(dev.get_element_descriptor())
  AHCI      SGPIO Enclosure   2.00
Element Descriptor diagnostic page:
  generation code: 0x0
  element descriptor list (grouped by type):
    Element type: Array device slot, subenclosure id: 0 [ti=0]
      Element type code: 23, Overall descriptor: Drive Slots
      Element 0 descriptor: Slot 00
      Element 1 descriptor: Slot 01
```
Enclosure Status page:
```
>>> print(dev.get_enclosure_status())
  AHCI      SGPIO Enclosure   2.00
Enclosure Status diagnostic page:
  INVOP=0, INFO=0, NON-CRIT=0, CRIT=0, UNRECOV=0
  generation code: 0x0
  status descriptor list
    Element type: Array device slot, subenclosure id: 0 [ti=0]
      Overall descriptor:
        00 00 00 00
      Element 0 descriptor:
        01 00 00 00
      Element 1 descriptor:
        01 00 00 00
```

To set the control, index and command should be passed as strings to `set_control()`. Index can be mentioned with or without type index (the first part), like `sg_ses`. If type index is not given, 0 is assumed. `set`, `get` and `clear` commands are supported.
```
>>> dev.set_control("0,1", "get=fault")
0
>>> dev.set_control("0,1", "set=fault")
0
>>> dev.set_control("0,1", "get=fault")
1
>>> dev.set_control("0,1", "clear=fault")
0
>>> dev.set_control("0,1", "get=fault")
0
>>> dev.set_control("0,1", "set=ident")
0
>>> dev.set_control("0,1", "get=ident")
1
```

### CODE REVIEW

File: ses/ses.pyx [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
+        cfg = self.get_configuration()
+        element_desc = self.get_element_descriptor()
+        enc_status = self.get_enclosure_status()
+
+        enclosure = {
+                "id": "",
+                "name": "",
+                "status": set(),
+                "elements": {},
+            }
```
Comment (yocalebo) [inline ses/ses.pyx]:
```suggestion
        }
```

**Review by amotin — COMMENTED** [review]
I'll leave review of the python code to Caleb.  But it is quite surprising to see you first generating full text output mimicking sg_ses and then parsing it back in your own code.  I understand the generation of the original output was done to ease middleware integration, and that is good to have right now for quick integration.  But in a long run it is not right.  I'm sorry that this code will need to be rewritten significantly once more at some point.  We should not use regular expressions to parse text that was just generated when it goes about 1200 disks every few minutes.

**Comment by umsaleem** [comment]
I agree. I think I can re-write `status()` add it later without parsing the human readable output provided by existing functions.

As @yocalebo suggests, we should initially keep the output that mimics `sg_ses`. I'll update the PR.

**Comment by umsaleem** [comment]
I have removed the `status()` for now.

File: ses/ses.pyx [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+                    desc_len = libsgutils.sg_get_unaligned_be16(bp + 2) + 4
+                    if desc_len > 4:
+                        self.start += snprintf(self.start, self.end - self.start, "      Element %d descriptor: %.*s\n", j, desc_len - 4, bp + 4)
+                    else:
+                        self.start += snprintf(self.start, self.end - self.start, "      Element %d descriptor: <empty>\n", j)
+                    bp += desc_len
+                tp += 1
+            self.clear_objs()
+            with gil:
+                    return bytes(self.r_buff, encoding='ascii').decode()
```
Comment (yocalebo) [inline ses/ses.pyx]:
```suggestion
                return bytes(self.r_buff, encoding='ascii').decode()
```

File: ses/ses.pyx [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+
+from libc.string cimport memset, memcpy, memcmp, strlen
+from libc.stdlib cimport free
+from libc.stdio cimport snprintf
+from libc.errno cimport errno
+from libc.stdint cimport uint8_t, uint32_t, uint64_t
+
+from enum import IntEnum
+from os import strerror
+import re
```
Comment (yocalebo) [inline ses/ses.pyx]:
```suggestion

```

File: ses/ses.pyx [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+        self.tmp_buff = libsgutils.sg_memalign(self.MX_ALLOC_LEN, 0, &self.free_tmp_buff, False)
+        if self.tmp_buff == NULL:
+            return -1
+        memset(self.tmp_buff, 0, self.MX_ALLOC_LEN)
+        return 0
+
+    cdef void free_resp_buffs(self) nogil:
+        if self.free_rsp_buff != NULL:
+            free(self.free_rsp_buff)
+        if self.free_rsp_buff != NULL:
```
Comment (yocalebo) [inline ses/ses.pyx]: We're core dumping here. I presume it should be 
```suggestion
        if self.free_tmp_buff != NULL:
```

**Review by yocalebo — CHANGES_REQUESTED** [review]

**Comment by umsaleem** [comment]
I have added the `status()` method that does not parse the human readable output provided by existing functions. Regular Expressions are not used. Diagnostic pages are parsed directly in `status()` to get the relevant information from the diagnostic pages and build the dictionary.

While testing on M60, there was a segmentation fault. It was because the return buffer (r_buff) was too small to hold the large text output that is to be returned when there are lots of disks. The size of return buffer is updated to 256 KB to fix this.

**Comment by amotin** [comment]
I see you are using snprintf's, etc, but still you are saying you've got segmentation a fault.  It sounds wrong.  And while bigger buffer may indeed be needed, just size increase does not sounds good, since it does not fix anything, only hides the problem for a while

**Comment by umsaleem** [comment]
Sure, I'll take another look at it and debug it further. The segmentation fault was occurring while trying to free the response buffers from `clear_resp_buff`, it did not occur while doing `snprintf`.

**Comment by umsaleem** [comment]
I have investigated and fixed the segmentation fault issue. `self.end - self.start` returns remaining size of `self.r_buff`, which is passed to `snprintf`. When `self.r_buff` runs out of space, this value becomes negative. `snprintf` treats this as unsigned value, and performs the write operation that causes the segmentation fault to occur.

To fix this, we check if `self.end -self.start` is greater than 0 before doing `snprintf`, and raise the error to indicate the buffer is full, if this is the case to avoid the segmentation fault.

I have tested this by reducing the size of self.r_buff. The segmentation fault does not occur now, instead the error is raised.

**Review by yocalebo — APPROVED** [review]
Tested and confirmed that it works as designed. This will allow easy integration for Cobia and work on using the "status" function for the next major release (which will require essentially a rewrite like we did on CORE 13+)

**Pull request merged**